src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp

```diff
@@ -15,751 +15,878 @@
+        }
+    }
 
-                    if (Creature* Brundir = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_BRUNDIR)))
-                        if (Brundir->IsAlive())
-                            Brundir->AI()->DoAction(ACTION_SUPERCHARGE);
+    void JustDied(Unit* /*killer*/) override
+    {
+        if (_events.IsInPhase(PHASE_THREE))
+        {
+            AOIEncounterHelper::CleanupAuras(me, _instance);
+            _instance->SetBossState(DATA_ASSEMBLY_OF_IRON, DONE);
+            DoCastAOE(SPELL_KILL_CREDIT, true);
+            Talk(SAY_STEELBREAKER_ENCOUNTER_DEFEATED);
+        }
+        else
+        {
+            me->SetLootRecipient(nullptr);
+            Talk(SAY_STEELBREAKER_DEATH);
+            DoCastAOE(SPELL_SUPERCHARGE);
+        }
+    }
 
-                    if (Creature* Molgeim = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_MOLGEIM)))
-                        if (Molgeim->IsAlive())
-                            Molgeim->AI()->DoAction(ACTION_SUPERCHARGE);
-                }
-            }
+    uint32 GetData(uint32 type) const override
+    {
+        if (type == DATA_PHASE_3)
+            return uint8(_events.IsInPhase(PHASE_THREE));
 
-            void KilledUnit(Unit* who) override
-            {
-                if (who->GetTypeId() == TYPEID_PLAYER)
-                    Talk(SAY_STEELBREAKER_SLAY);
+        return 0;
+    }
 
-                if (phase == 3)
-                    DoCast(me, SPELL_ELECTRICAL_CHARGE);
-            }
+    void DoAction(int32 action) override
+    {
+        switch (action)
+        {
+            case ACTION_SUPERCHARGE:
+                _events.SetPhase(!_events.IsInPhase(PHASE_TWO) ? PHASE_TWO : PHASE_THREE);
+                _events.RescheduleEvent(EVENT_FUSION_PUNCH, 13s, 22s);
+
+                if (_events.IsInPhase(PHASE_TWO) || _events.IsInPhase(PHASE_THREE))
+                    _events.RescheduleEvent(EVENT_STATIC_DISRUPTION, 20s, 40s);
+
+                if (_events.IsInPhase(PHASE_THREE))
+                    _events.RescheduleEvent(EVENT_OVERWHELMING_POWER, 2s, 5s);
+                break;
+            case ACTION_ADD_CHARGE:
+                if (_events.IsInPhase(PHASE_THREE))
+                    DoCastSelf(SPELL_ELECTRICAL_CHARGE, true);
+                break;
+            case ACTION_BERSERK_TRIGGERED:
+                Talk(SAY_STEELBREAKER_BERSERK);
+                break;
+            default:
+                break;
+        }
+    }
 
-            void UpdateAI(uint32 diff) override
-            {
-                if (!UpdateVictim())
-                    return;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim() || !CheckInRoom())
+            return;
 
-                events.Update(diff);
+        _events.Update(diff);
 
-                if (me->HasUnitState(UNIT_STATE_CASTING))
-                    return;
+        if (me->HasUnitState(UNIT_STATE_CASTING) || me->HasSpellFocus())
```

Comment:
HasSpellFocus check can be removed after this commit?https://github.com/TrinityCore/TrinityCore/commit/684910a230b2e8e780ad24bc0c4eece835bc249a

Reply:
No. The additional spell focus check here is to make sure that the boss wont launch events in quick succession.

Reply:
@Ovahlord accept me in discord, sucker <3